### PR TITLE
Alteia data source

### PR DESCRIPTION
**Pull request by MuchosChuchos**
implement data source that parses an HTML-page

### CODE REVIEW

File: api/build.gradle [hunk 1]
```diff
@@ -46,6 +46,7 @@ dependencies {
     implementation 'org.flywaydb:flyway-core:7.15.0'
     implementation 'io.springfox:springfox-boot-starter:3.0.0'
     implementation 'org.modelmapper:modelmapper:2.4.4'
+	implementation 'org.jsoup:jsoup:1.14.3'
```
Comment (Oleksandr-Eleks) [inline api/build.gradle:49]: Please format properly

File: api/src/main/resources/application.yaml [hunk 2]
```diff
@@ -32,3 +32,7 @@ pharmagator:
       category-id: !!str 4628621
       medicament-category-id: !!str 4628712
       pharmacy-name: apteka-rozetka
+    apteka-alteia:
+      url: https://apteka.alteia.ua/shop
+      products-fetch-url: /page
+      pharmacy-name: apteka-alteia
```
Comment (Oleksandr-Eleks) [inline api/src/main/resources/application.yaml:38]: Add BL in the end of file

File: api/src/main/resources/db/migration/V1_0_3__change_externalId_length.sql [hunk 3]
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE prices
+    ALTER COLUMN external_id SET DATA TYPE varchar(200);
```
Comment (Oleksandr-Eleks) [inline api/src/main/resources/db/migration/V1_0_3__change_externalId_length.sql:2]: Add BL in the end of file

Comment (Oleksandr-Eleks) [inline api/src/main/resources/db/migration/V1_0_3__change_externalId_length.sql:2]: Make it 255

File: api/src/main/resources/db/migration/V1_0_3__change_externalId_length.sql [hunk 4]
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE prices
```
Comment (Oleksandr-Eleks) [inline api/src/main/resources/db/migration/V1_0_3__change_externalId_length.sql:1]: Wrong script version

File: api/src/main/java/com/eleks/academy/pharmagator/dataproviders/PharmacyAlteiaDataProvider.java [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+        Stream<MedicineDto> medicineDtoStream = this.getMedicineStreamFromPage(page);
+
+        int lastPage;
+        int currentPage = 2;
+
+        do {
+            Document newPage = Jsoup.parse(this.fetchHtmlByPage(currentPage));
+            Elements pageNumbers = newPage.select("ul[class=page-numbers]").select("a[class=page-numbers]");
+            lastPage = Integer.parseInt(pageNumbers.last().text());
```
Comment (Oleksandr-Eleks) [inline api/src/main/java/com/eleks/academy/pharmagator/dataproviders/PharmacyAlteiaDataProvider.java:50]: Is it always the same value?
I mean, if so, why do You need to recalculate it each iteration?

File: api/src/main/java/com/eleks/academy/pharmagator/dataproviders/PharmacyAlteiaDataProvider.java [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+
+            medicineDtoStream = Stream.concat(medicineDtoStream, this.getMedicineStreamFromPage(newPage));
+            currentPage++;
+        } while (currentPage <= lastPage);
+
+        return medicineDtoStream;
+    }
+
+    private Stream<MedicineDto> getMedicineStreamFromPage(Document page) {
```
Comment (Oleksandr-Eleks) [inline api/src/main/java/com/eleks/academy/pharmagator/dataproviders/PharmacyAlteiaDataProvider.java:59]: Useless method

File: api/src/main/java/com/eleks/academy/pharmagator/dataproviders/PharmacyAlteiaDataProvider.java [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+        return this.getMedicineFromHtml(page).stream();
+    }
+
+    private List<MedicineDto> getMedicineFromHtml(Document page) {
+        List<MedicineDto> medicineDtos = new ArrayList<>();
+        Elements medsName = page.select("h2[class=woocommerce-loop-product__title]");
+        Element products = page.select("ul[class=products columns-4]").first();
+        Elements medsPrice = products.select("span[class=woocommerce-Price-amount amount]");
+        Elements medsId = page.select("a[class=woocommerce-LoopProduct-link woocommerce-loop-product__link]");
+        for (int i = 0; i < medsName.size(); i++) {
```
Comment (Oleksandr-Eleks) [inline api/src/main/java/com/eleks/academy/pharmagator/dataproviders/PharmacyAlteiaDataProvider.java:69]: Stream?

File: api/src/main/java/com/eleks/academy/pharmagator/dataproviders/PharmacyAlteiaDataProvider.java [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+            if ("".equals(medsName.get(i).text())) {
```
Comment (Oleksandr-Eleks) [inline api/src/main/java/com/eleks/academy/pharmagator/dataproviders/PharmacyAlteiaDataProvider.java:70]: `.trim().isEmpty()`

File: api/src/main/java/com/eleks/academy/pharmagator/dataproviders/PharmacyAlteiaDataProvider.java [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+        return this.getMedicineDtoStreamFromHtml(html);
+    }
+
+    private Stream<MedicineDto> getMedicineDtoStreamFromHtml(String html) {
+        Document page = Jsoup.parse(html);
+
+        Stream<MedicineDto> medicineDtoStream = this.getMedicineStreamFromPage(page);
+
+        int lastPage;
+        int currentPage = 2;
```
Comment (Oleksandr-Eleks) [inline api/src/main/java/com/eleks/academy/pharmagator/dataproviders/PharmacyAlteiaDataProvider.java:45]: Magic number?
Extract to config.

File: api/src/main/java/com/eleks/academy/pharmagator/dataproviders/PharmacyAlteiaDataProvider.java [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+            if ("".equals(medsName.get(i).text())) {
+                medsName.remove(i);
+                medsId.remove(i);
+            }
+            try {
+                medicineDtos.add(MedicineDto.builder()
+                        .externalId(this.getTextBetween(medsId.get(i).attr("href"), "product/", "/"))
+                        .price(new BigDecimal(this.getTextBetween(medsPrice.get(i).text(), "", "\u20B4").replaceAll(",", "")))
+                        .title(medsName.get(i).text())
+                        .pharmacyName("alteia")
```
Comment (Oleksandr-Eleks) [inline api/src/main/java/com/eleks/academy/pharmagator/dataproviders/PharmacyAlteiaDataProvider.java:79]: You should use 'pharmacy-name' from config here

File: api/src/main/java/com/eleks/academy/pharmagator/dataproviders/PharmacyAlteiaDataProvider.java (hunk 5 above)
Comment (MuchosChuchos) [reply api/src/main/java/com/eleks/academy/pharmagator/dataproviders/PharmacyAlteiaDataProvider.java:50]: Not as you expect. Each request takes about 1min, hence the number of pages may be changed. I've tested it.

File: api/build.gradle (hunk 1 above)
Comment (MuchosChuchos) [reply api/build.gradle:49]: it was out of my view

File: api/src/main/java/com/eleks/academy/pharmagator/dataproviders/PharmacyAlteiaDataProvider.java (hunk 6 above)
Comment (MuchosChuchos) [reply api/src/main/java/com/eleks/academy/pharmagator/dataproviders/PharmacyAlteiaDataProvider.java:59]: I agree. It was for Uncle Bob=)

**Pull request closed without merging**